File: 0NotCompleted/Union-Find/UnionFind.java
public class UnionFind{

    private int[] id;
    private int[] size;
    private int count;

    public UnionFind(int n) {
        id = new int[n];
        size = new int[n];
        count = n;
        for(int i = 0; i<n;i++){
            id[i]=i;
            size[i]=1;
        }
    }

    public int find(int parent){
        int searchNumber = parent;
        while(parent != id[parent]){
            parent = id[parent];
        }
        compress(searchNumber, parent);
        return parent;
    }

    private void compress(int parent, int root) {
        while(parent != id[parent]){
            id[parent] = root;
            parent = id[parent];
        }
    }

    public boolean query(int s, int t){
        return find(s) == find(t);
    }

    public void union(int s, int t) {
        int rootS = find(s);
        int rootT = find(t);

        if(rootS != rootT){
            count--;
            if(size[rootS]<size[rootT]){
                id[rootS] = rootT; 
                size[rootT] += size[rootS];}
            else{
                id[rootT] = rootS; 
                size[rootS] += size[rootT];}
        } 
    }

    public int getCount() {
        return count;
    }

}

File: Tildes/UnionFind.py
class UnionFind:

    def __init__(self, n):
        self.components = n
        self.id = list(range(n))
        self.size = [1]*n

    def find(self, parent):
        baseNum = parent
        while(parent != self.id[parent]):
            parent = self.id[parent]

        self.pathCompress(baseNum, parent)
        return parent

    def pathCompress(self, parent, root):
        while(parent != self.id[parent]):
            nextParent = self.id[parent]
            self.id[parent] = root
            parent = nextParent

    def query(self,s,t):
        return self.find(s) == self.find(t)
    
    def union(self,s,t):
        rootS = self.find(s)
        rootT = self.find(t)

        if(rootS == rootT): 
            return

        self.components -= 1
        if(self.size[rootS] < self.size[rootT]):
            self.id[rootS] = self.id[rootT]
            self.size[rootT] += self.size[rootS]
        else:
            self.id[rootT] = self.id[rootS]
            self.size[rootS] += self.size[rootT]

    def getComponents(self):
        return self.components

    def getSize(self,group):
        return self.size[self.find(group)]


